### PR TITLE
#1307 Integration test for Streaming Conformance

### LINKED ISSUE #1307: Add an end-to-end Kafka source&sink streaming conformance integration test
## Background
An end-to-end streaming conformance test will help to catch streaming-only issues.
Such as the Catalyst optimizer issue #1306  

## Feature
Add an end-to-end Kafka source&sink streaming conformance integration test.

* Data for the test can be usual JSONs. We can even reuse the existing end-to-end conformance test data for that.
* An instance of Kafka should be brought up.
* Data from the examples should be sent to raw Kafka topic converting the input data to Avro.
* A streaming conformance test should be run as Kafka -> Parquet and Kadka -> Kafka

## Proposed Solution [Optional]
Use https://www.testcontainers.org/modules/kafka/
It is used by Hyperdrive as well. Thanks Felipe for the suggestion.

### PR DESCRIPTION
**Pull request by AdrianOlosutean**
Closes #1307

### CODE REVIEW

**Comment by AdrianOlosutean** [comment]
So far I've addded a test case with timeout which can catch #1306 . What other use cases could be useful for testing HyperConformance?

File: spark-jobs/src/main/scala/za/co/absa/enceladus/conformance/HyperConformance.scala [hunk 1]
```diff
@@ -40,29 +41,30 @@ class HyperConformance (implicit cmd: ConfCmdConfig,
...
 
-    val reportVersion = getReportVersion
-
     logPreConformanceInfo(rawDf)
 
-    val infoDateColumn = infoDateFactory.getInfoDateColumn(rawDf)
-
     val conformance = dao.getDataset(cmd.datasetName, cmd.datasetVersion)
+    val conformedDf = applyConformanceTransformations(rawDf, conformance)
+    conformedDf
```
Comment (yruslan) [inline spark-jobs/src/main/scala/za/co/absa/enceladus/conformance/HyperConformance.scala:54]: Why the logging was removed. Maybe let's put it here.

File: spark-jobs/src/test/scala/za/co/absa/enceladus/conformance/streaming/HyperConformanceIntegrationSuite.scala [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+import za.co.absa.enceladus.conformance.interpreter.fixtures.{NestedStructsFixture, StreamingFixture}
+
+class HyperConformanceIntegrationSuite extends FunSuite with StreamingFixture with NestedStructsFixture {
+
+  test("Test with catalyst workaround") {
+    val frame: DataFrame = testHyperConformance(standardizedDf,
+      "result",
+      nestedStructsDS)
+    frame.show()
+    assertResult(frame.count())(20)
```
Comment (yruslan) [inline spark-jobs/src/test/scala/za/co/absa/enceladus/conformance/streaming/HyperConformanceIntegrationSuite.scala:29]: In addition to testing count, maybe we can test the output as well?

File: spark-jobs/src/test/scala/za/co/absa/enceladus/conformance/interpreter/fixtures/StreamingFixture.scala [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+import za.co.absa.enceladus.conformance.streaming.{InfoDateFactory, InfoDateLiteralFactory}
+import za.co.absa.enceladus.dao.MenasDAO
+import za.co.absa.enceladus.model.Dataset
+import za.co.absa.enceladus.utils.testUtils.SparkTestBase
+
+trait StreamingFixture extends FunSuite with SparkTestBase with MockitoSugar {
+  implicit val menasBaseUrls: List[String] = List()
+  implicit val cmd: ConfCmdConfig = ConfCmdConfig.apply(reportVersion = Some(1))
+  implicit val infoDateFactory: InfoDateFactory = new InfoDateLiteralFactory("2020-05-23")
+  private val interval = 10000
```
Comment (yruslan) [inline spark-jobs/src/test/scala/za/co/absa/enceladus/conformance/interpreter/fixtures/StreamingFixture.scala:35]: How this interval was chosen?

Comment (AdrianOlosutean) [reply spark-jobs/src/test/scala/za/co/absa/enceladus/conformance/interpreter/fixtures/StreamingFixture.scala:35]: Mostly I lowered the interval as long as it was possible to reproduce the bug. Since on the test where the optimizer workaround is enabled, it works to write the results during this interval, but on the other one it doesn't. From testing with several values I noticed not enabling this interval would keep the test stuck, so adding the timeout made sense.

**Comment by AdrianOlosutean** [comment]
1. Yes, I was able to reproduce it by enabling or disabling the workaround and running it on the test. As @kevinwallimann said before, this bug is happening even if you provide just a MemoryStream.
2. The class seems to take around 30 seconds

File: spark-jobs/src/test/scala/za/co/absa/enceladus/conformance/interpreter/fixtures/StreamingFixture.scala [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    val sink = conformed
+      .writeStream
+      .trigger(Trigger.ProcessingTime(interval))
+      .queryName(memoryTableName)
+      .outputMode("append")
+      .format("memory")
+      .start()
+
+    input.collect().foreach(e => memoryStream.addData(e))
+    sink.awaitTermination(interval)
```
Comment (kevinwallimann) [inline spark-jobs/src/test/scala/za/co/absa/enceladus/conformance/interpreter/fixtures/StreamingFixture.scala:60]: `awaitTermination(interval)` is probably a little bit dangerous. If the query is terminated just before the microbatch has been committed, then the whole microbatch will be aborted and there will be zero rows.
Also, you have to assume that your data will be processed within `interval` ms. 10s is probably enough, but it's kind of a magic value.
In this case `processAllAvailable` would be a good fit, because then you can be sure that all the rows will have been read, no matter to what value `interval` has been set to. `processAllAvailable` terminates as long as the data source doesn't emit values indefinitely which is the case here

Comment (AdrianOlosutean) [reply spark-jobs/src/test/scala/za/co/absa/enceladus/conformance/interpreter/fixtures/StreamingFixture.scala:60]: I agreee, but how can you use `processAllAvailable` on a test with the catalyst optimizer workaround disabled without having an unreasonable running time for testing?

Comment (kevinwallimann) [reply spark-jobs/src/test/scala/za/co/absa/enceladus/conformance/interpreter/fixtures/StreamingFixture.scala:60]: Ok, now I see your need for a timeout. Either you give up on testing with multiple micro-batches and use `Trigger.Once` or you could use a margin of safety, say `3 * interval`.
Also, if you really want to cover multiple micro-batches, you need to make sure that the `foreach` loop runs during multiple micro-batches.

File: spark-jobs/src/test/scala/za/co/absa/enceladus/conformance/streaming/HyperConformanceIntegrationSuite.scala [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+import org.scalatest.FunSuite
+import za.co.absa.enceladus.conformance.interpreter.fixtures.{NestedStructsFixture, StreamingFixture}
+
+class HyperConformanceIntegrationSuite extends FunSuite with StreamingFixture with NestedStructsFixture {
+
+  test("Test with catalyst workaround") {
+    val frame: DataFrame = testHyperConformance(standardizedDf,
+      "result",
+      nestedStructsDS)
+    frame.show()
```
Comment (Zejnilovic) [inline spark-jobs/src/test/scala/za/co/absa/enceladus/conformance/streaming/HyperConformanceIntegrationSuite.scala:29]: Don't use show in tests

File: spark-jobs/src/test/scala/za/co/absa/enceladus/conformance/streaming/HyperConformanceIntegrationSuite.scala [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+    val strings = data("strings").asInstanceOf[GenericRowWithSchema]
+    assertResult(strings.getAs("whitespaces_upper"))("K  C H E   RS Z     ")
+
+    val numerics = data("numerics").asInstanceOf[GenericRowWithSchema]
+    assertResult(numerics.getAs("big_negative_negated"))(783143645497786L)
+
+    assertResult(frame.count())(20)
+  }
+
+  test("Test without catalyst workaround") {
```
Comment (Zejnilovic) [inline spark-jobs/src/test/scala/za/co/absa/enceladus/conformance/streaming/HyperConformanceIntegrationSuite.scala:50]: I am not sure what this tests?

File: spark-jobs/src/test/scala/za/co/absa/enceladus/conformance/interpreter/fixtures/StreamingFixture.scala [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+      .trigger(Trigger.ProcessingTime(processingTime))
+      .queryName(sinkTableName)
+      .outputMode("append")
+      .format("memory")
+      .start()
+
+    input.collect().foreach(e => {
+      memoryStream.addData(e)
+      Thread.sleep(100)
+      spark.sql(s"select * from $sinkTableName").show()
```
Comment (yruslan) [inline spark-jobs/src/test/scala/za/co/absa/enceladus/conformance/interpreter/fixtures/StreamingFixture.scala:62]: `.show` is redundant here since we don't want our build logs to be very big.

File: spark-jobs/src/test/scala/za/co/absa/enceladus/conformance/interpreter/fixtures/StreamingFixture.scala [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+import org.scalatest.mockito.MockitoSugar
+import za.co.absa.enceladus.conformance.{ConfCmdConfig, HyperConformance}
+import za.co.absa.enceladus.conformance.interpreter.FeatureSwitches
+import za.co.absa.enceladus.conformance.streaming.{InfoDateFactory, InfoDateLiteralFactory}
+import za.co.absa.enceladus.dao.MenasDAO
+import za.co.absa.enceladus.model.Dataset
+import za.co.absa.enceladus.utils.testUtils.SparkTestBase
+
+trait StreamingFixture extends FunSuite with SparkTestBase with MockitoSugar {
+  implicit val menasBaseUrls: List[String] = List()
```
Comment (Zejnilovic) [inline spark-jobs/src/test/scala/za/co/absa/enceladus/conformance/interpreter/fixtures/StreamingFixture.scala:32]: I know it is a matter of taste, but readability wise I like `List.empty[String]` the most.

File: spark-jobs/src/test/scala/za/co/absa/enceladus/conformance/interpreter/fixtures/StreamingFixture.scala [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+      .writeStream
+      .trigger(Trigger.ProcessingTime(processingTime))
+      .queryName(sinkTableName)
+      .outputMode("append")
+      .format("memory")
+      .start()
+
+    input.collect().foreach(e => {
+      memoryStream.addData(e)
+      Thread.sleep(100)
```
Comment (yruslan) [inline spark-jobs/src/test/scala/za/co/absa/enceladus/conformance/interpreter/fixtures/StreamingFixture.scala:61]: Is there a way it can be done so it won't need to rely on timing. Since if the timing is involved we might see the test fail non-deterministically since it is going to be run on machines that have very different specs.

Maybe you can wait for a micro-batch to be processed before sending another one by doing `select` and `count` on the memory table?

File: spark-jobs/src/test/scala/za/co/absa/enceladus/conformance/streaming/HyperConformanceIntegrationSuite.scala [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+
+import org.apache.spark.sql.DataFrame
+import org.apache.spark.sql.catalyst.expressions.GenericRowWithSchema
+import org.scalatest.FunSuite
+import za.co.absa.enceladus.conformance.interpreter.fixtures.{NestedStructsFixture, StreamingFixture}
+
+class HyperConformanceIntegrationSuite extends FunSuite with StreamingFixture with NestedStructsFixture {
+
+  test("Test with catalyst workaround") {
+    val frame: DataFrame = testHyperConformance(standardizedDf,
```
Comment (yruslan) [inline spark-jobs/src/test/scala/za/co/absa/enceladus/conformance/streaming/HyperConformanceIntegrationSuite.scala:26]: We name data frames `df` everywhere. I'm not arguing which one is better, just let's be consistent as much as possible.

File: spark-jobs/src/test/scala/za/co/absa/enceladus/conformance/interpreter/fixtures/NestedStructsFixture.scala [hunk 11]
```diff
@@ -141,7 +138,6 @@ trait NestedStructsFixture extends BeforeAndAfterAll with SparkTestBase {
 
   protected val nestedStructsCastingDS = Dataset(
     name = "Nested Structs Conformance",
-    version = 1,
```
Comment (Zejnilovic) [inline spark-jobs/src/test/scala/za/co/absa/enceladus/conformance/interpreter/fixtures/NestedStructsFixture.scala:144]: Why was this removed?

File: spark-jobs/src/test/scala/za/co/absa/enceladus/conformance/streaming/HyperConformanceIntegrationSuite.scala [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+      "result",
+      nestedStructsDS)
+    frame.show()
+    val fields = frame.schema.fieldNames
+    val data = frame.first().getValuesMap(fields)
```
Comment (yruslan) [inline spark-jobs/src/test/scala/za/co/absa/enceladus/conformance/streaming/HyperConformanceIntegrationSuite.scala:31]: Rather than comparing specific values, we can compare full output by doing:
```
df.toJSON.collect().mkString("\n")
```
and comparing results with the expected JSON somewhere in resources. If there. are non-deterministic columns, we can remove them from comparison by using `dropColumn()` or `select()`.

I'd prefer doing `select` on all fields expected in the output data frame. Since changes of column order are allowed, doing `select` will make th order specific for easier comparison. 

Ideally, the expected output JSON should be the same for batch and streaming jobs. If the order of records could be changes, I'd recommend doing `orderBy` before comparison.

File: spark-jobs/src/test/scala/za/co/absa/enceladus/conformance/streaming/HyperConformanceIntegrationSuite.scala [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+
+    val infoDateStringColumnName = "enceladus_info_date_string"
+    assert(fields.contains(infoDateStringColumnName))
+    assertResult(data("enceladus_info_date_string"))("2020-05-23")
```
Comment (yruslan) [inline spark-jobs/src/test/scala/za/co/absa/enceladus/conformance/streaming/HyperConformanceIntegrationSuite.scala:35]: We need to test different `enceladus_info_date` strategies, possibly as part of smaller tests.

File: spark-jobs/src/test/scala/za/co/absa/enceladus/conformance/streaming/HyperConformanceIntegrationSuite.scala [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+    assert(fields.contains(infoVersionColumnName))
+    assertResult(data(infoVersionColumnName))(1)
+
+    val strings = data("strings").asInstanceOf[GenericRowWithSchema]
+    assertResult(strings.getAs("whitespaces_upper"))("K  C H E   RS Z     ")
+
+    val numerics = data("numerics").asInstanceOf[GenericRowWithSchema]
+    assertResult(numerics.getAs("big_negative_negated"))(783143645497786L)
+
+    assertResult(frame.count())(20)
```
Comment (yruslan) [inline spark-jobs/src/test/scala/za/co/absa/enceladus/conformance/streaming/HyperConformanceIntegrationSuite.scala:47]: Let's make count checking the first test. If it fails it doesn't make sense to compare individual values or the output as a whole.

File: spark-jobs/src/test/scala/za/co/absa/enceladus/conformance/streaming/HyperConformanceIntegrationSuite.scala [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+  }
+
+  test("Test without catalyst workaround") {
+    val frame: DataFrame = testHyperConformance(standardizedDf,
```
Comment (yruslan) [inline spark-jobs/src/test/scala/za/co/absa/enceladus/conformance/streaming/HyperConformanceIntegrationSuite.scala:51]: What does this test?

File: spark-jobs/src/test/scala/za/co/absa/enceladus/conformance/streaming/HyperConformanceIntegrationSuite.scala [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+      "result2",
+      nestedStructsDS,
+      catalystWorkaround = false)
+    frame.show()
```
Comment (yruslan) [inline spark-jobs/src/test/scala/za/co/absa/enceladus/conformance/streaming/HyperConformanceIntegrationSuite.scala:55]: Redundant show

File: spark-jobs/src/test/scala/za/co/absa/enceladus/conformance/interpreter/fixtures/NestedStructsFixture.scala (hunk 11 above)
Comment (AdrianOlosutean) [reply spark-jobs/src/test/scala/za/co/absa/enceladus/conformance/interpreter/fixtures/NestedStructsFixture.scala:144]: It's redundant, the default is 1

File: spark-jobs/src/test/scala/za/co/absa/enceladus/conformance/streaming/HyperConformanceIntegrationSuite.scala (hunk 5 above)
Comment (AdrianOlosutean) [reply spark-jobs/src/test/scala/za/co/absa/enceladus/conformance/streaming/HyperConformanceIntegrationSuite.scala:29]: I mostly thought that show would make following what happens at the runtime easier

File: spark-jobs/src/test/scala/za/co/absa/enceladus/conformance/streaming/HyperConformanceIntegrationSuite.scala (hunk 6 above)
Comment (AdrianOlosutean) [reply spark-jobs/src/test/scala/za/co/absa/enceladus/conformance/streaming/HyperConformanceIntegrationSuite.scala:50]: The idea is that since the workaround is disabled, the test will timeout and no data will be added to the sink in a reasonable time. Probably makes sense to add more description

File: spark-jobs/src/test/scala/za/co/absa/enceladus/conformance/interpreter/fixtures/StreamingFixture.scala (hunk 7 above)
Comment (AdrianOlosutean) [reply spark-jobs/src/test/scala/za/co/absa/enceladus/conformance/interpreter/fixtures/StreamingFixture.scala:62]: Would printing the length of the table make sense instead for debugging reasons?

File: spark-jobs/src/test/scala/za/co/absa/enceladus/conformance/interpreter/fixtures/StreamingFixture.scala (hunk 9 above)
Comment (AdrianOlosutean) [reply spark-jobs/src/test/scala/za/co/absa/enceladus/conformance/interpreter/fixtures/StreamingFixture.scala:61]: About timing, I don't say the tests couldn't fail, but it's not likely. As @kevinwallimann suggested, the timeout between the insertion better simulates a more realistic use case and since inserting everything would take 2 seconds, the total waited time is 6 seconds and I can incrase it. 

The other discussion we had about not using `awaitTermination` and using `processAllAvailable` is that using the second one makes testing with the workaround disabled run an unreasonable amount of time.

File: spark-jobs/src/test/scala/za/co/absa/enceladus/conformance/streaming/HyperConformanceIntegrationSuite.scala (hunk 12 above)
Comment (AdrianOlosutean) [reply spark-jobs/src/test/scala/za/co/absa/enceladus/conformance/streaming/HyperConformanceIntegrationSuite.scala:31]: Makes sense

File: spark-jobs/src/test/scala/za/co/absa/enceladus/conformance/streaming/HyperConformanceIntegrationSuite.scala (hunk 6 above)
Comment (Zejnilovic) [reply spark-jobs/src/test/scala/za/co/absa/enceladus/conformance/streaming/HyperConformanceIntegrationSuite.scala:50]: This is testing of the presence of a bug is Spark. Seems quite weird to me and I would remove the test.

File: spark-jobs/src/test/scala/za/co/absa/enceladus/conformance/streaming/HyperConformanceIntegrationSuite.scala (hunk 5 above)
Comment (Zejnilovic) [reply spark-jobs/src/test/scala/za/co/absa/enceladus/conformance/streaming/HyperConformanceIntegrationSuite.scala:29]: I understand, but in an ideal world, you just look at the bottom where it says that everything passed. In case of an issue, every one of us can write this line by themselves and re-run

File: spark-jobs/src/test/scala/za/co/absa/enceladus/conformance/interpreter/fixtures/StreamingFixture.scala (hunk 7 above)
Comment (yruslan) [reply spark-jobs/src/test/scala/za/co/absa/enceladus/conformance/interpreter/fixtures/StreamingFixture.scala:62]: You can use `log.debug` if you really want to commit lines that are used only for debugging. But generally, we prefer removing lines added for debugging but instead, provide a detailed error message if the test fails.

File: spark-jobs/src/test/scala/za/co/absa/enceladus/conformance/streaming/HyperConformanceIntegrationSuite.scala (hunk 15 above)
Comment (yruslan) [reply spark-jobs/src/test/scala/za/co/absa/enceladus/conformance/streaming/HyperConformanceIntegrationSuite.scala:51]: I see Saša has the same question :)

File: spark-jobs/src/test/scala/za/co/absa/enceladus/conformance/interpreter/fixtures/StreamingFixture.scala (hunk 9 above)
Comment (yruslan) [reply spark-jobs/src/test/scala/za/co/absa/enceladus/conformance/interpreter/fixtures/StreamingFixture.scala:61]: Sorry, I wasn't very constructive since I don't have a solution to this. Structured Streaming jobs is a new topic for me as well.
Not sure if it can be done without relying on the timer. But if it could be done, such a solution would be preferable.

File: spark-jobs/src/test/scala/za/co/absa/enceladus/conformance/streaming/HyperConformanceIntegrationSuite.scala (hunk 6 above)
Comment (yruslan) [reply spark-jobs/src/test/scala/za/co/absa/enceladus/conformance/streaming/HyperConformanceIntegrationSuite.scala:50]: It might make sense to keep a commented test that, once uncommented, will demonstrate that the Catalyst issue still exists and the test hangs.
It might be useful in case new version of Spark has this issue fixed. We could try it then to see if the fix covers our use case.

File: spark-jobs/src/test/scala/za/co/absa/enceladus/conformance/interpreter/fixtures/StreamingFixture.scala (hunk 4 above)
Comment (yruslan) [reply spark-jobs/src/test/scala/za/co/absa/enceladus/conformance/interpreter/fixtures/StreamingFixture.scala]: I'm slightly confused. We need to test 2 things:
1. The correctness of the results of streaming conformance. All workarounds should be turned on so such tests can finish in a reasonable time.
2. We need to test that workaround works by adding a test that hangs if the workaround is turned off. There is no particular time limit on the test, but should be some reasonable time.

Tests that test (1) should not rely on timing if possible. If `processAllAvailable` helps with it - great..
Tests that test (2) might rely on timing, but not necessarily. Currently in batch, if the workaround stops working the corresponding test will just hang forever. Not ideal, but we expect workaround to work. If it stops having an effect, it will be noticed by the developer when the test hangs.
Alternatively, a big timeout can be used to interrupt the test if it takes too long. But the value of the timeout should be set so that test passes on all envs the test could run, including Jenkins VMs.

The reason I care about this so much is that timings introduce non-determinism. Such tests could work perfectly locally on Mac, but start failing on Windows or on Linux build box.

File: spark-jobs/src/test/scala/za/co/absa/enceladus/conformance/interpreter/fixtures/StreamingFixture.scala (hunk 9 above)
Comment (yruslan) [reply spark-jobs/src/test/scala/za/co/absa/enceladus/conformance/interpreter/fixtures/StreamingFixture.scala:61]: Thanks a lot, @kevinwallimann for the insight and for the proposed solution!

@AdrianOlosutean, I'm still not 100% sure if we should have a [uncommented] test that runs forever if the workaround is turned off and is interrupted by a timer. Let me know what you think. My current concern is that having such a test would bring more risks than benefits. We need to ensure that the processing interval does interrupt the query even if it is stuck inside the Catalyst layer.
Having a commented test that hangs without the workaround can be beneficial to test new versions of Spark. And since it is commented, it doesn't take time to run on normal builds :)

Comment (AdrianOlosutean) [reply spark-jobs/src/test/scala/za/co/absa/enceladus/conformance/interpreter/fixtures/StreamingFixture.scala:61]: For me it seems to that the commented test would be the more sensible approach. Very interesting, obviously the solution proposed by @kevinwallimann is the rigurous one and I guess it should be used, thanks.

**Comment by AdrianOlosutean** [comment]
Release notes: [here](https://github.com/absaoss/enceladus/issues/1307#issuecomment-644697488)
